lib/std/os/linux.zig

```diff
@@ -6816,6 +6816,139 @@ pub const ifreq = extern struct {
     },
 };
 
+pub const PACKET = struct {
+    pub const HOST = 0;
+    pub const BROADCAST = 1;
+    pub const MULTICAST = 2;
+    pub const OTHERHOST = 3;
+    pub const OUTGOING = 4;
+    pub const LOOPBACK = 5;
+    pub const USER = 6;
+    pub const KERNEL = 7;
+    pub const FASTROUTE = 8;
+
+    pub const ADD_MEMBERSHIP = 1;
+    pub const DROP_MEMBERSHIP = 2;
+    pub const RECV_OUTPUT = 3;
+    pub const RX_RING = 5;
+    pub const STATISTICS = 6;
+    pub const COPY_THRESH = 7;
+    pub const AUXDATA = 8;
+    pub const ORIGDEV = 9;
+    pub const VERSION = 10;
+    pub const HDRLEN = 11;
+    pub const RESERVE = 12;
+    pub const TX_RING = 13;
+    pub const LOSS = 14;
+    pub const VNET_HDR = 15;
+    pub const TX_TIMESTAMP = 16;
+    pub const TIMESTAMP = 17;
+    pub const FANOUT = 18;
+    pub const TX_HAS_OFF = 19;
+    pub const QDISC_BYPASS = 20;
+    pub const ROLLOVER_STATS = 21;
+    pub const FANOUT_DATA = 22;
+    pub const IGNORE_OUTGOING = 23;
+    pub const VNET_HDR_SZ = 24;
+
+    pub const FANOUT_HASH = 0;
+    pub const FANOUT_LB = 1;
+    pub const FANOUT_CPU = 2;
+    pub const FANOUT_ROLLOVER = 3;
+    pub const FANOUT_RND = 4;
+    pub const FANOUT_QM = 5;
+    pub const FANOUT_CBPF = 6;
+    pub const FANOUT_EBPF = 7;
+    pub const FANOUT_FLAG_ROLLOVER = 0x1000;
+    pub const FANOUT_FLAG_UNIQUEID = 0x2000;
+    pub const FANOUT_FLAG_IGNORE_OUTGOING = 0x4000;
+    pub const FANOUT_FLAG_DEFRAG = 0x8000;
+};
+
+pub const tpacket_versions = enum(u32) {
+    V1 = 0,
+    V2 = 1,
+    V3 = 2,
+};
+
+pub const tpacket_req3 = extern struct {
+    block_size: c_uint, // Minimal size of contiguous block
+    block_nr: c_uint, // Number of blocks
+    frame_size: c_uint, // Size of frame
+    frame_nr: c_uint, // Total number of frames
+    retire_blk_tov: c_uint, // Timeout in msecs
+    sizeof_priv: c_uint, // Offset to private data area
+    feature_req_word: c_uint,
+};
+
+pub const tpacket_bd_ts = extern struct {
+    sec: c_uint,
+    frac: extern union {
+        usec: c_uint,
+        nsec: c_uint,
+    },
+};
+
+pub const TP_STATUS = extern union {
+    rx: packed struct(u32) {
+        USER: bool,
+        COPY: bool,
+        LOSING: bool,
+        CSUMNOTREADY: bool,
+        VLAN_VALID: bool,
+        BLK_TMO: bool,
+        VLAN_TPID_VALID: bool,
+        CSUM_VALID: bool,
+        GSO_TCP: bool,
+        _: u20,
+        TS_SOFTWARE: bool,
+        TS_SYS_HARDWARE: bool,
+        TS_RAW_HARDWARE: bool,
+    },
+    tx: packed struct(u32) {
+        SEND_REQUEST: bool,
+        SENDING: bool,
+        WRONG_FORMAT: bool,
+        _: u26,
+        TS_SOFTWARE: bool,
+        TS_SYS_HARDWARE: bool,
+        TS_RAW_HARDWARE: bool,
+    },
+};
+
+pub const tpacket_block_desc = extern struct {
+    version: u32,
+    offset_to_priv: u32,
+    block_status: TP_STATUS,
```

Comment:
What's the reason for inlining `tpacket_bd_header_u` here?

Reply:
Just less cruft. It's a union with 1 member that's only referenced there, so I thought I'd just add it to the base struct.

Presumably it'd be useful if the API ever gets extended, but that hasn't happened in more than a decade so I didn't bother.

Reply:
My concern here is ABI compatibility. If it was a `packed struct`, there'd definitely be no issue, but for an `extern struct` you'd have to be able to know for sure that all architectures/C ABIs that the kernel supports consider the inlining of `tpacket_bd_header_u` to be equivalent to declaring a field of type `tpacket_bd_header_u`. (I don't personally know the answer to that.)

Reply:
The union should inherit the struct's size and alignment. The struct then seems densely packed since it's forcibly 8-byte aligned, 8 bytes into the parent struct, a multiple of 8 bytes in size and doesn't need padding.

If we're still unsure I can restore the struct/union wrapping though.

Reply:
I think you're right that this is fine from an ABI perspective.

However, regardless of that, I think we should use a union here. The reason a union is used in the C API seems to be so that the header layout can be changed from `tpacket_hdr_v1` to a theoretical `tpacket_hdr_v2` without breaking backwards compatibility.

If we don't use a union and mirror the C API here then such a theoretical addition of a union field would require likely an API breaking change on the Zig side, which I think is less than ideal.

Reply:
> a theoretical addition of a union field would require likely an API breaking change on the Zig side, which I think is less than ideal.

I just figured since this hasn't happened in 12 years it won't happen in the future... But yes, it'd for sure be annoying if it does. I'll go ahead and fix it.